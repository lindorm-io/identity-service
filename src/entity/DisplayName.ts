import { DisplayNameEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { includes } from "lodash";

export interface IDisplayName extends IEntity {
  name: string;
  numbers: Array<number>;
}

export interface IDisplayNameOptions extends IEntityBaseOptions {
  name: string;
  numbers?: Array<number>;
}

export class DisplayName extends EntityBase {
  private _name: string;
  private _numbers: Array<number>;

  constructor(options: IDisplayNameOptions) {
    super(options);
    this._name = options.name;
    this._numbers = options.numbers || [];
  }

  public get name(): string {
    return this._name;
  }

  public get numbers(): Array<number> {
    return this._numbers;
  }

  public add(number: number): void {
    if (includes(this._numbers, number)) {
      throw new Error("Number already exists for this DisplayName");
    }

    this._numbers.push(number);
    this.addEvent(DisplayNameEvent.NUMBER_ADDED, { number });
  }

  public remove(number: number): void {
    const array: Array<number> = [];

    for (const item of this._numbers) {
      if (item === number) continue;
      array.push(item);
    }

    this._numbers = array;
    this.addEvent(DisplayNameEvent.NUMBER_REMOVED, { number });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DisplayNameEvent.CREATED) continue;
      throw new EntityCreationError("DisplayName");
    }

    this.addEvent(DisplayNameEvent.CREATED, {
      name: this._name,
      numbers: this._numbers,
    });
  }
}
