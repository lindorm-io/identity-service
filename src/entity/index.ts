export * from "./DisplayName";
export * from "./Identity";
