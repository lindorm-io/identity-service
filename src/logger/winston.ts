import { Logger, LogLevel } from "@lindorm-io/winston";
import { NODE_ENVIRONMENT } from "../config";
import { NodeEnvironment } from "@lindorm-io/core";
import { join } from "path";
import { readFileSync } from "fs";
import { sanitiseToken } from "@lindorm-io/jwt";

const pkg = readFileSync(join(__dirname, "..", "..", "package.json"), { encoding: "utf8" });
const { name, version } = JSON.parse(pkg);

export const winston = new Logger({
  packageName: name,
  packageVersion: version,
  test: NODE_ENVIRONMENT === NodeEnvironment.TEST,
});

if (NODE_ENVIRONMENT === NodeEnvironment.PRODUCTION) {
  winston.addFileTransport(LogLevel.ERROR);
  winston.addFileTransport(LogLevel.WARN);
  winston.addFileTransport(LogLevel.INFO);
  winston.addFileTransport(LogLevel.VERBOSE);
  winston.addFileTransport(LogLevel.DEBUG);
  winston.addFileTransport(LogLevel.SILLY);
}

if (NODE_ENVIRONMENT !== NodeEnvironment.PRODUCTION) {
  winston.addConsole(LogLevel.DEBUG);
}

winston.addFilter("request.header.authorization", sanitiseToken);
