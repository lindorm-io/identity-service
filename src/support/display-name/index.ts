export * from "./find-random-number";
export * from "./get-display-name";
export * from "./remove-display-name-number";
