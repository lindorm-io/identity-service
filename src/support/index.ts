export * from "./display-name";
export * from "./identity";
export * from "./token";
