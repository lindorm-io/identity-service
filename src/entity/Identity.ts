import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { IdentityEvent } from "../enum";

export interface IIdentityAddress {
  country: string;
  locality: string;
  postalCode: string;
  region: string;
  streetAddress: string;
}

export interface IIdentityDisplayName {
  name: string;
  number: number;
}

export interface IIdentity extends IEntity {
  // public
  displayName: IIdentityDisplayName;
  gravatar: string;

  // private
  username: string;

  // openid
  address: IIdentityAddress;
  birthDate: string;
  familyName: string;
  gender: string;
  givenName: string;
  locale: string;
  middleName: string;
  nickname: string;
  phoneNumber: string;
  phoneNumberVerified: boolean;
  picture: string;
  preferredUsername: string;
  profile: string;
  website: string;
  zoneInfo: string;
}

export interface IIdentityOptions extends IEntityBaseOptions {
  // public
  displayName?: IIdentityDisplayName;
  gravatar?: string;

  // private
  username?: string;

  // openid
  address?: IIdentityAddress;
  birthDate?: string;
  familyName?: string;
  gender?: string;
  givenName?: string;
  locale?: string;
  middleName?: string;
  nickname?: string;
  phoneNumber?: string;
  phoneNumberVerified?: boolean;
  picture?: string;
  preferredUsername?: string;
  profile?: string;
  website?: string;
  zoneInfo?: string;
}

export class Identity extends EntityBase {
  private _address: IIdentityAddress;
  private _birthDate: string;
  private _displayName: IIdentityDisplayName;
  private _familyName: string;
  private _gender: string;
  private _givenName: string;
  private _gravatar: string;
  private _locale: string;
  private _middleName: string;
  private _nickname: string;
  private _phoneNumber: string;
  private _phoneNumberVerified: boolean;
  private _picture: string;
  private _preferredUsername: string;
  private _profile: string;
  private _username: string;
  private _website: string;
  private _zoneInfo: string;

  constructor(options: IIdentityOptions) {
    super(options);
    this._address = options.address || null;
    this._birthDate = options.birthDate || null;
    this._displayName = options.displayName || null;
    this._familyName = options.familyName || null;
    this._gender = options.gender || null;
    this._givenName = options.givenName || null;
    this._gravatar = options.gravatar || null;
    this._locale = options.locale || null;
    this._middleName = options.middleName || null;
    this._nickname = options.nickname || null;
    this._phoneNumber = options.phoneNumber || null;
    this._phoneNumberVerified = options.phoneNumberVerified || false;
    this._picture = options.picture || null;
    this._preferredUsername = options.preferredUsername || null;
    this._profile = options.profile || null;
    this._username = options.username || null;
    this._website = options.website || null;
    this._zoneInfo = options.zoneInfo || null;
  }

  public get address(): IIdentityAddress {
    return this._address;
  }
  public set address(address: IIdentityAddress) {
    this._address = address;
    this.addEvent(IdentityEvent.ADDRESS_CHANGED, { address: this._address });
  }

  public get birthDate(): string {
    return this._birthDate;
  }
  public set birthDate(birthDate: string) {
    this._birthDate = birthDate;
    this.addEvent(IdentityEvent.BIRTH_DATE_CHANGED, { birthDate: this._birthDate });
  }

  public get displayName(): IIdentityDisplayName {
    return this._displayName;
  }
  public set displayName(displayName: IIdentityDisplayName) {
    this._displayName = displayName;
    this.addEvent(IdentityEvent.DISPLAY_NAME_CHANGED, { displayName: this._displayName });
  }

  public get familyName(): string {
    return this._familyName;
  }
  public set familyName(familyName: string) {
    this._familyName = familyName;
    this.addEvent(IdentityEvent.FAMILY_NAME_CHANGED, { familyName: this._familyName });
  }

  public get gender(): string {
    return this._gender;
  }
  public set gender(gender: string) {
    this._gender = gender;
    this.addEvent(IdentityEvent.GENDER_CHANGED, { gender: this._gender });
  }

  public get givenName(): string {
    return this._givenName;
  }
  public set givenName(givenName: string) {
    this._givenName = givenName;
    this.addEvent(IdentityEvent.GIVEN_NAME_CHANGED, { givenName: this._givenName });
  }

  public get gravatar(): string {
    return this._gravatar;
  }
  public set gravatar(gravatar: string) {
    this._gravatar = gravatar;
    this.addEvent(IdentityEvent.GRAVATAR_CHANGED, { gravatar: this._givenName });
  }

  public get locale(): string {
    return this._locale;
  }
  public set locale(locale: string) {
    this._locale = locale;
    this.addEvent(IdentityEvent.LOCALE_CHANGED, { locale: this._locale });
  }

  public get middleName(): string {
    return this._middleName;
  }
  public set middleName(middleName: string) {
    this._middleName = middleName;
    this.addEvent(IdentityEvent.MIDDLE_NAME_CHANGED, { middleName: this._middleName });
  }

  public get nickname(): string {
    return this._nickname;
  }
  public set nickname(nickname: string) {
    this._nickname = nickname;
    this.addEvent(IdentityEvent.NICKNAME_CHANGED, { nickname: this._nickname });
  }

  public get phoneNumber(): string {
    return this._phoneNumber;
  }
  public set phoneNumber(phoneNumber: string) {
    this._phoneNumber = phoneNumber;
    this.addEvent(IdentityEvent.PHONE_NUMBER_CHANGED, { phoneNumber: this._phoneNumber });
  }

  public get phoneNumberVerified(): boolean {
    return this._phoneNumberVerified;
  }
  public set phoneNumberVerified(phoneNumberVerified: boolean) {
    this._phoneNumberVerified = phoneNumberVerified;
    this.addEvent(IdentityEvent.PHONE_NUMBER_VERIFIED_CHANGED, { phoneNumberVerified: this._phoneNumber });
  }

  public get picture(): string {
    return this._picture;
  }
  public set picture(picture: string) {
    this._picture = picture;
    this.addEvent(IdentityEvent.PICTURE_CHANGED, { picture: this._picture });
  }

  public get preferredUsername(): string {
    return this._preferredUsername;
  }
  public set preferredUsername(preferredUsername: string) {
    this._preferredUsername = preferredUsername;
    this.addEvent(IdentityEvent.PREFERRED_USERNAME_CHANGED, { preferredUsername: this._preferredUsername });
  }

  public get profile(): string {
    return this._profile;
  }
  public set profile(profile: string) {
    this._profile = profile;
    this.addEvent(IdentityEvent.PROFILE_CHANGED, { profile: this._profile });
  }

  public get username(): string {
    return this._username;
  }
  public set username(username: string) {
    this._username = username;
    this.addEvent(IdentityEvent.USERNAME_CHANGED, { username: this._profile });
  }

  public get website(): string {
    return this._website;
  }
  public set website(website: string) {
    this._website = website;
    this.addEvent(IdentityEvent.WEBSITE_CHANGED, { website: this._website });
  }

  public get zoneInfo(): string {
    return this._zoneInfo;
  }
  public set zoneInfo(zoneInfo: string) {
    this._zoneInfo = zoneInfo;
    this.addEvent(IdentityEvent.ZONE_INFO_CHANGED, { zoneInfo: this._zoneInfo });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== IdentityEvent.CREATED) continue;
      throw new EntityCreationError("Identity");
    }

    this.addEvent(IdentityEvent.CREATED, {
      address: this._address,
      birthDate: this._birthDate,
      displayName: this._displayName,
      familyName: this._familyName,
      gender: this._gender,
      givenName: this._givenName,
      gravatar: this._gravatar,
      locale: this._locale,
      middleName: this._middleName,
      nickname: this._nickname,
      phoneNumber: this._phoneNumber,
      phoneNumberVerified: this._phoneNumberVerified,
      picture: this._picture,
      preferredUsername: this._preferredUsername,
      profile: this._profile,
      username: this._username,
      website: this._website,
      zoneInfo: this._zoneInfo,
    });
  }
}
