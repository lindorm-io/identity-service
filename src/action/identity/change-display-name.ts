import Joi from "@hapi/joi";
import { IKoaIdentityContext } from "../../typing";
import { assertAccountPermission, getDisplayNameObject, removeDisplayNameNumber } from "../../support";
import { isScope, Scope } from "@lindorm-io/jwt";
import { InvalidScopeError } from "../../error";

export interface IChangeDisplayNameOptions {
  identityId: string;
  displayName: string;
}

const schema = Joi.object({
  identityId: Joi.string().guid().required(),
  displayName: Joi.string().required(),
});

export const changeDisplayName = (ctx: IKoaIdentityContext) => async (
  options: IChangeDisplayNameOptions,
): Promise<void> => {
  await schema.validateAsync(options);

  const { logger, repository, token } = ctx;
  const {
    bearer: { scope },
  } = token;
  const { identityId, displayName } = options;

  await assertAccountPermission(ctx)(identityId);

  if (!isScope(scope, Scope.EDIT)) {
    throw new InvalidScopeError(scope, Scope.EDIT);
  }

  const identity = await repository.identity.find({ id: identityId });
  const displayNameExists = !!identity.displayName;

  identity.displayName = await getDisplayNameObject(ctx)(displayName);

  await repository.identity.update(identity);

  if (displayNameExists) {
    await removeDisplayNameNumber(ctx)(identity.displayName);
  }

  logger.debug("identity display name updated", {
    identityId,
    displayName: identity.displayName,
  });
};
