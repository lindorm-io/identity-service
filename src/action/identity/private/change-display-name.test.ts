import MockDate from "mockdate";
import { Identity } from "../../../entity";
import { Scope } from "@lindorm-io/jwt";
import { changeDisplayName } from "./change-display-name";
import { getGreyBoxRepository, inMemoryStore, resetStore } from "../../../test";
import { removeDisplayNameNumber } from "../../../support";
import { winston } from "../../../logger";

jest.mock("../../../support", () => ({
  assertAccountPermission: jest.fn(() => () => undefined),
  getDisplayNameObject: jest.fn(() => (name: string) => ({
    name,
    number: 9999,
  })),
  removeDisplayNameNumber: jest.fn(() => () => undefined),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("changeDisplayName", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      logger: winston,
      repository: await getGreyBoxRepository(),
      token: { bearer: { scope: [Scope.DEFAULT, Scope.EDIT].join(" ") } },
    };
  });

  afterEach(resetStore);

  test("should add new display name", async () => {
    await ctx.repository.identity.create(
      new Identity({
        id: "fd4e3548-1279-4065-9671-74c3bbea0c25",
      }),
    );

    const options = {
      identityId: "fd4e3548-1279-4065-9671-74c3bbea0c25",
      displayName: "newName",
    };

    await expect(changeDisplayName(ctx)(options)).resolves.toBe(undefined);

    expect(inMemoryStore).toMatchSnapshot();
    expect(removeDisplayNameNumber).not.toHaveBeenCalled();
  });

  test("should change existing display name", async () => {
    await ctx.repository.identity.create(
      new Identity({
        id: "fd4e3548-1279-4065-9671-74c3bbea0c25",
        displayName: {
          name: "displayName",
          number: 1111,
        },
      }),
    );

    const options = {
      identityId: "fd4e3548-1279-4065-9671-74c3bbea0c25",
      displayName: "newName",
    };

    await expect(changeDisplayName(ctx)(options)).resolves.toBe(undefined);

    expect(inMemoryStore).toMatchSnapshot();
    expect(removeDisplayNameNumber).toHaveBeenCalled();
  });
});
