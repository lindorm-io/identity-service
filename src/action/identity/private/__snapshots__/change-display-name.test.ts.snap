// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeDisplayName should add new display name 1`] = `
Object {
  "identity": Object {
    "identity": Object {
      "data": Array [
        Object {
          "address": null,
          "birthDate": null,
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": Object {
            "name": "newName",
            "number": 9999,
          },
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": null,
                "birthDate": null,
                "displayName": null,
                "familyName": null,
                "gender": null,
                "givenName": null,
                "gravatar": null,
                "locale": null,
                "middleName": null,
                "nickname": null,
                "phoneNumber": null,
                "phoneNumberVerified": false,
                "picture": null,
                "preferredUsername": null,
                "profile": null,
                "username": null,
                "website": null,
                "zoneInfo": null,
              },
            },
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_display_name_changed",
              "payload": Object {
                "displayName": Object {
                  "name": "newName",
                  "number": 9999,
                },
              },
            },
          ],
          "familyName": null,
          "gender": null,
          "givenName": null,
          "gravatar": null,
          "id": "fd4e3548-1279-4065-9671-74c3bbea0c25",
          "locale": null,
          "middleName": null,
          "nickname": null,
          "phoneNumber": null,
          "phoneNumberVerified": false,
          "picture": null,
          "preferredUsername": null,
          "profile": null,
          "updated": 2020-01-01T07:00:00.000Z,
          "username": null,
          "version": 1,
          "website": null,
          "zoneInfo": null,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;

exports[`changeDisplayName should change existing display name 1`] = `
Object {
  "identity": Object {
    "identity": Object {
      "data": Array [
        Object {
          "address": null,
          "birthDate": null,
          "created": 2020-01-01T07:00:00.000Z,
          "displayName": Object {
            "name": "newName",
            "number": 9999,
          },
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_created",
              "payload": Object {
                "address": null,
                "birthDate": null,
                "displayName": Object {
                  "name": "displayName",
                  "number": 1111,
                },
                "familyName": null,
                "gender": null,
                "givenName": null,
                "gravatar": null,
                "locale": null,
                "middleName": null,
                "nickname": null,
                "phoneNumber": null,
                "phoneNumberVerified": false,
                "picture": null,
                "preferredUsername": null,
                "profile": null,
                "username": null,
                "website": null,
                "zoneInfo": null,
              },
            },
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "identity_display_name_changed",
              "payload": Object {
                "displayName": Object {
                  "name": "newName",
                  "number": 9999,
                },
              },
            },
          ],
          "familyName": null,
          "gender": null,
          "givenName": null,
          "gravatar": null,
          "id": "fd4e3548-1279-4065-9671-74c3bbea0c25",
          "locale": null,
          "middleName": null,
          "nickname": null,
          "phoneNumber": null,
          "phoneNumberVerified": false,
          "picture": null,
          "preferredUsername": null,
          "profile": null,
          "updated": 2020-01-01T07:00:00.000Z,
          "username": null,
          "version": 1,
          "website": null,
          "zoneInfo": null,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
